lib/ansible/modules/crypto/openssl_cert.py

```diff
@@ -0,0 +1,996 @@
+        # Implemented in child classes
+        pass
+
+    def remove(self):
+        '''Remove the certificate'''
+
+        try:
+            os.remove(self.dest)
+        except OSError:
+            self.changed = False
+
+    def dump(self):
+        '''Serialize the object into a dictionary.'''
+        # Implemented in child classes
+        pass
+
+    def check(self):
+        '''Check properties of the generated certificate.'''
+
+        if not os.path.exists(self.dest):
+            self.module.fail_json(msg='The certificate at %s does not exist' % self.dest)
+
+        # pyOpenSSL checks:
+        with open(self.dest, "r") as certfile:
+            cert = crypto.load_certificate(crypto.FILETYPE_PEM, certfile.read())
+
+        if self.signature_algorithms is not None:
+            if cert.get_signature_algorithm() not in self.signature_algorithms:
+                self.module.fail_json(
+                    msg='Invalid signature algorithm (got %s, expected one of %s)' % (cert.get_signature_algorithm(), self.signature_algorithms))
+
+        if self.subject is not None:
+            cert_subject = cert.get_subject().get_components()
+            for entry in self.subject:
+                if (entry, self.subject[entry]) not in cert_subject:
+                    self.module.fail_json(
+                        msg='Invalid subject component (got %s, expected all of %s to be present)' % (cert_subject, self.subject))
+
+        if self.issuer is not None:
+            cert_issuer = cert.get_issuer().get_components()
+            for entry in self.issuer:
+                if (entry, self.issuer[entry]) not in cert_issuer:
+                    self.module.fail_json(
+                        msg='Invalid issuer component (got %s, expected all of %s to be present)' % (cert_issuer, self.issuer))
+
+        if self.has_expired is not None:
+            expired_check = cert.has_expired()
+            if self.has_expired != expired_check:
+                self.module.fail_json(
+                    msg='Certificate expiration check failed (certificate expiration is %s, expected %s)' % (expired_check, self.has_expired))
+
+        if self.version is not None:
+            # Version numbers in certs are off by one:
+            # v1: 0, v2: 1, v3: 2 ...
+            cert_version = cert.get_version() + 1
+            if self.version != cert_version:
+                self.module.fail_json(
+                    msg='Invalid certificate version number (got %s, expected %s)' % (cert_version, self.version))
+
+        if self.privatekey_path is not None:
+            with open(self.privatekey_path, "r") as private_key_file:
+                privkey = crypto.load_privatekey(crypto.FILETYPE_PEM, private_key_file.read())
+            try:
+                pubkey = privkey.to_cryptography_key().public_key()
+                cert_pubkey = cert.get_pubkey().to_cryptography_key()
+                if pubkey != cert_pubkey:
```

Comment:
The comparison doesn't work even for valid values.

I've run the following commands:

  * `openssl x509 -noout -modulus -in m.ansible.com.crt | openssl md5`
  * `openssl rsa -noout -modulus -in m.ansible.com.pem | openssl md5`
  * `openssl req -noout -modulus -in m.ansible.com.csr | openssl md5`

They all output the same md5 meaning they're all valid for each other yet this test fails.

I'd guess the content of the object needs to be compared, not the object themselves